src/Microsoft.Diagnostics.Monitoring.EventPipe/Triggers/Pipelines/TraceEventTriggerPipeline.cs

```diff
@@ -0,0 +1,111 @@
+        {
+            _callback = callback ?? throw new ArgumentNullException(nameof(callback));
+            _eventSource = eventSource ?? throw new ArgumentNullException(nameof(eventSource));
+            _trigger = trigger ?? throw new ArgumentNullException(nameof(trigger));
+
+            IDictionary<string, IEnumerable<string>> providerEventMap = _trigger.GetProviderEventMap();
+            if (null == providerEventMap)
+            {
+                // Allow all events to be forwarded to the trigger
+                _eventSource.Dynamic.AddCallbackForProviderEvents(
+                    null,
+                    TraceEventCallback);
+            }
+            else
+            {
+                // Only allow events described in the mapping to be forwarded to the trigger.
+                // If a provider has no events specified, then all events from that provider are forwarded.
+                _eventSource.Dynamic.AddCallbackForProviderEvents(
+                    (string providerName, string eventName) =>
+                    {
+                        if (!providerEventMap.TryGetValue(providerName, out IEnumerable<string> eventNames))
+                        {
+                            return EventFilterResponse.RejectProvider;
+                        }
+                        else if (null == eventNames)
+                        {
+                            return EventFilterResponse.AcceptEvent;
+                        }
+                        else if (!eventNames.Contains(eventName, StringComparer.Ordinal))
+                        {
+                            return EventFilterResponse.RejectEvent;
+                        }
+                        return EventFilterResponse.AcceptEvent;
+                    },
+                    TraceEventCallback);
+            }
+        }
+
+        protected override async Task OnRun(CancellationToken token)
+        {
+            using var _ = token.Register(() => _completionSource.TrySetCanceled(token));
+
+            await _completionSource.Task.ConfigureAwait(false);
+        }
+
+        protected override Task OnStop(CancellationToken token)
+        {
+            _completionSource.TrySetResult(null);
+
+            return base.OnStop(token);
+        }
+
+        protected override Task OnCleanup()
+        {
+            _completionSource.TrySetCanceled();
+
+            _eventSource.Dynamic.RemoveCallback<TraceEvent>(TraceEventCallback);
+
+            return base.OnCleanup();
+        }
+
+        private void TraceEventCallback(TraceEvent obj)
+        {
+            // Check if processing of in-flight events should be ignored
+            // due to pipeline in the midst of stopping.
+            if (!_completionSource.Task.IsCompleted)
+            {
+                // If the trigger condition has been satified, invoke the callback
+                if (_trigger.HasSatisfiedCondition(obj))
+                {
+                    _callback(obj);
```

Comment:
So if the callback is very time consuming, it may cause a lot of backpressure on handling events on the provider. Do we want to force this to be processed asynchronously rather than pushing that decision onto the callee?

Reply:
There's a couple of scenarios to think about:
- If you were to load a nettrace file into an `EventPipeEventSource` and feed that directly into `TraceEventTriggerPipeline` instances, you'd probably want this callback to be synchronous to the event so that way you can observe that the trigger condition holds for successive events. If the callback were async, you'd get a bunch of callbacks for each event and your handler would likely be running its code in parallel for many of those callbacks, which is probably not desirable. Maybe the current callback API is still not sufficient to convey the notion the successive events hold the condition; maybe the callback needs to indicate that the first invocation is the first event to hold the condition and then the callback is not invoked again until the last event in the successive observation of events that holds the condition occurs. Maybe @delmyers or @WilliamXieMSFT have some feedback on how this might be used in detecting interesting conditions in an offline trace.
- For a "live" event source, your suggestion makes more sense. However, it still has the problem of successive events that might hold the condition would invoke the callback on different threads and the callbacks might execute in parallel. We might want a mode where the trigger is suspended/disabled after invoking the callback and must be explicitly resumed/enabled when the caller is interested in detecting the condition again. I have plans for this type of behavior in a subsequent change that was briefly described in the PR description.

Without going down the route of async callbacks, with the current implementation, callers should be able to stop the pipeline during the callback in order to prevent future successive callbacks and to flush out the event pipe buffer.

I'll keep thinking about how to balance some of these different requirements.

Reply:
If there isn't a requirement that triggers are synchronous within `dotnet-monitor`, then theoretically you could queue the callbacks to be invoked orthogonally to the stream processing. For now, a live session is already asynchronous with regards to when a listener receives the events compared to when it was sent.

Reply:
> If there isn't a requirement that triggers are synchronous within `dotnet-monitor`, then theoretically you could queue the callbacks to be invoked orthogonally to the stream processing. For now, a live session is already asynchronous with regards to when a listener receives the events compared to when it was sent.

I would contend that the consumer of the EventPipeTriggerPipeline should decide if its wants to implement that queue functionality or just respond to the events directly within the callbacks.

For `dotnet-monitor`, I believe that the callback will end up just invoking `TaskCompletionSource<T>.TrySetResult`, which shouldn't be anything significant in terms of compute time, especially if the task completion is run asynchronously (pass `TaskCreationOptions.RunContinuationsAsynchronously` in the TCS ctor).